pennylane/operation.py

```diff
@@ -763,6 +770,51 @@ class DiagonalOperation(Operation):
     """
     # pylint: disable=abstract-method
 
+    @classmethod
+    def _eigvals(cls, *params):
+        """Eigenvalues of the operator.
+
+        The order of the eigenvalues needs to match the order of
+        the computational basis vectors.
+
+        This is a *class method* that must be defined for all
+        new diagonal operations, that returns the eigenvalues
+        of the operator in the computational basis.
+
+        This private method allows eigenvalues to be computed
+        directly without instantiating the operators first.
+
+        To return the eigenvalues of *instantiated* operators,
+        please use the :attr:`~.Operator.eigvals` property instead.
+
+        **Example:**
+
+        >>> qml.RZ._eigvals(0.5)
+        >>> array([0.96891242-0.24740396j, 0.96891242+0.24740396j])
+
+        Returns:
+            array: eigenvalue representation
+        """
+        raise NotImplementedError
+
+    @property
+    def eigvals(self):
+        r"""Eigenvalues of an instantiated diagonal operation.
+
+        The order of the eigenvalues needs matches the order of
+        the computational basis vectors.
+
+        **Example:**
+
+        >>> U = qml.RZ(0.5, wires=1)
```

Comment:
Might it be confusing to users if they look at the docstring of, say, `Hadamard.eigvals` and find an example for `RZ`?

Reply:
It might. But this is also the case for `matrix`. Doing it differently would require us to overwrite the property for every class.